#!/bin/bash
set -e
# load programs needed
module load fastqc
# [+] Loading fastqc 0.11.9
module load cutadapt
# cutadapt/4.0
module load STAR
# [+] Loading STAR  2.7.10b
module load samtools
# [+] Loading samtools 1.17  ...
module load stringtie
# [+] Loading stringtie  2.2.1
module load deeptools
# [+] Loading stringtie  3.5.1

###### Get sample_folder full path for 1 library from command line
# folder with 2 fastq files for 1 library
# NOTE: run the script in the directory containing the sample_folder.
#       then use the sample_folder as the relative path: short one_word path
sample_folder=$1 # get from the command line. # Use short relative path!!!
                 # Ony 1 folder for 1 library with forward and reverse fastq files
				 
				 
######### Parameters related to STAR ###############
# STAR genome index is generated in TruSeq_totalRNA-Seq.sh script
# genome_STAR_index_Dir="/data/guang/genome_index/RNA-STAR/RNA_STAR_overhang100/Mus_musculus.GRCm38.97"
genome_STAR_index_Dir="/data/yangg3/mouse_genome_index/RNA_STAR_GRCm38.101_with_rDNA_Overhang99"
####################################################

######### Parameter for StringTie #############
gtf_full_annotation="/data/yangg3/mouse_genome_index/RNA_STAR_GRCm38.101_with_rDNA_Overhang99/Mus_musculus.GRCm38.101_plus_rDNA.gtf"
# full_annotation include all genes and rDNA

###############################################
##******************* Parameter Definition Finished Here *********************##

##*************** The actual alignment script starts here. ********************##
################################################################################
cd $sample_folder
# Get into the sample_folder with fastq.gz file(s)


############# FastQC

mkdir -p $sample_folder\_fastqc_results
# Make a new folder in sample_folder to store FastQC result; Full path used here.
# The fastq files are stored as "fastq.gz" files. Check this !!!!
fastqc -t 8 *.fastq.gz -O ./$sample_folder\_fastqc_results/
# -t 8: use 8 threads
# relative path is used. The full path is 
# $sample_folder/$sample_folder\_fastqc_results
# Two files generate after calling fastqc: 



########################################################################################
################# Trimming 

mkdir -p $sample_folder\_cutadapt_trimmed
# Make a new folder in sample_folder to store trimmed result; Full path used here.
fastq_gz_file="$(ls *.fastq.gz)"

cutadapt --cores=8 --minimum-length 15 --cut 3 -a AAAAAAAAAA -o ./$sample_folder\_cutadapt_trimmed/$sample_folder.trimmed.fq.gz $fastq_gz_file

# --cores=8: use 8 thread
# --minimum-length 15: after trimming, only reads 15 bp or longer will be retained, while shorter ones
# will be discarded. Use -m 15 instead if a more stringent cutoff is desired.
# --cut 3: trim the first 3 bp of all reads. This takes care of removing extra bases inserted due to
# the SMART template-switching mechanism
#  -a AAAAAAAAAA: adapter sequence to be identified and removed. The adapter
# sequence and everything 3â€² of the adapter (including any Illumina adapter sequence) will
# be removed. By default, cutadapt will accept an error rate of 10%.

########################################################################################
################# RNA-STAR alignment 


mkdir -p $sample_folder\_RNA_STAR_Results_full_annotation

trimmed_fastq_gz_file="$(ls $PWD/$sample_folder\_cutadapt\_trimmed/*.fq.gz)"
#$PWD is current path
#This command get the full path of trimmed fastq.gz files

	
STAR --runThreadN 8 \
    --genomeDir $genome_STAR_index_Dir \
    --readFilesCommand zcat \
    --outSAMtype BAM SortedByCoordinate \
	--quantMode GeneCounts \
	--twopassMode Basic \
	--sjdbGTFfile $gtf_full_annotation \
    --readFilesIn $trimmed_fastq_gz_file \
    --outReadsUnmapped Fastx \
    --outFileNamePrefix ./$sample_folder\_RNA_STAR_Results_full_annotation/STAR_alignment\_$sample_folder\_

# index the bam file generated by RNA-STAR (already SortedByCoordinate)
samtools index $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/*_Aligned.sortedByCoord.out.bam

# get uniquely mapped reads
samtools view -hb -q 255 $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/*_Aligned.sortedByCoord.out.bam > $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/$sample_folder\_Aligned.sortedByCoord.uniq_mapped.bam 
samtools index $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/*_Aligned.sortedByCoord.uniq_mapped.bam

bamCoverage -b $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/$sample_folder\_Aligned.sortedByCoord.uniq_mapped.bam -o $PWD/$sample_folder\_RNA_STAR_Results_full_annotation/$sample_folder\_uniq_mapping_RPKM_normalized.bw --normalizeUsing RPKM --numberOfProcessors 8
##****************************Part2 Finished here *******************************##


##******** Part 3.1 Using StringTie to get gene expression table from merged .bam and .gtf files **********##

mkdir -p $sample_folder\_StringTie_Results_full_annotation

stringtie -p 8 -G $gtf_full_annotation \
    -B -e -o ./$sample_folder\_StringTie_Results_full_annotation/$sample_folder\_transcripts\_WITHeOption.gtf \
    -A ./$sample_folder\_StringTie_Results_full_annotation/$sample_folder\_gene_abundances.tsv \
    ./$sample_folder\_RNA_STAR_Results_full_annotation/*_Aligned.sortedByCoord.uniq_mapped.bam
    # The last parameter is the .bam file generated by RNA-STAR

##*********************** Part 3.1 Finished here ****************************************##
